### ISSUE TITLE
Edge cases require all implementations to couple authorization and storage

### ISSUE THREAD

**Issue by RubenVerborgh** [solid/web-access-control-spec#97]
## Context

For performance and scalability reasons, a desired property of an access control mechanism is that it can make independent decisions.

Specifically, if an access control system A is protecting an interface I, then we want to eliminate or at least reduce the number of requests from A to I.

Concretely, for Solid, it is undesirable for a request to depend on resource state, as this causes roundtrips between the WAC system and the LDP system.

## Issue

It used to be possible for Solid to build the WAC and LDP interfaces as separate systems/microservices, where the WAC stage completes before the LDP stage. WAC would be a gateway that, based on the request, either returned `403` or let the request through to the next stage. However, a couple of decisions make this independence impossible.

When looking at https://github.com/solid/specification/issues/14#issuecomment-683480525, all cases where the behavior for "exists"/"doesn't exist" have different `4xx` codes, the LDP backend to be accessed by the WAC system.

They are:

```
GET C/R
HEAD C/R
OPTIONS C/R
```

|C/|C/R|C/R exists|C/R doesn't exist|
|-|-|-|-|
|Read|Write|403|404|

```
PUT C/R
```

|C/|C/R|C/R exists|C/R doesn't exist|
|-|-|-|-|
|-|Write|200|403|

```
PATCH C/R
```
(not listed in the original comment, added by me here)
|C/|C/R|C/R exists|C/R doesn't exist|
|-|-|-|-|
|-|Write|200|403|

## Relevant spec text

The spec currently reads:

> When an operation requests to create a resource as a member of a container resource…

This specific "when" is expensive. A WAC system cannot tell for a given `PUT` or `PATCH` request whether it is a creation by looking at the request itself. It has to go to the backend and check whether that resource exists.

## Example

When a `PUT` request arrives for C/R, the required modes depend on the state of the LDP system. Hence, the WAC system needs to contact the LDP system to verify whether C/R exists before it knows what Authorization blocks to check for. The WAC system needs to put a lock on C/R in the LDP system for the duration of the request.

The situation gets worse when taking into account

> When a server supports the creation of intermediate containers in the process of creating a resource, the server applies the same requirements for the create operation for each container.

Meaning that, for every `PUT` request (!), we require a number of roundtrips up to the length of the path, making operations to `/a/b/c/d/e` potentially expensive.

## Potential solutions
Requiring the same permissions regardless of whether a resource exists.

Tagging @acoburn for implementation feedback, @kjetilk because we discussed an auth proxy in the past (which is currently no longer possible).

**Comment by RubenVerborgh** [comment]
Not to mention the possibility of a DDOS with `PUT /a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a`.

**Comment by csarven** [comment]
An agent needs Read on C/R for a server to return 200 (when resource exists) or 404 (when resource does not exist). How else are typical 200 or 404 are returned?

If agent does not have Read on C/R, the server returns 403. It doesn't have to do anything further. That'd be a conforming implementation.

If however, a server wants to return a more specific/useful code by taking in other information into account, eg. can the agent actually know C/R exists by viewing C/ if they have Read on C/? Those servers wanting to do more would still be conformant if they returned 404 (IMO). That's one of the reasons why there is a difference between the codes. https://github.com/solid/specification/issues/288#issuecomment-888102563 mentions they are not fixed for various reasons.

---

Don't we have some existing issues/discussions in Protocol on locking? Happy to revise the text in WAC so that it doesn't necessarily mean that WAC has to check with the system for each container or whatever. I'd expect the system to only query (like ASK) for Authorization rules that's relevant as opposed to for each path segment eg. if /a/ exists but /a/a/ doesn't exist, don't need to work from the lowest in path to highest (towards root).

**Comment by RubenVerborgh** [comment]
I'm afraid we're talking about different things; the cases you list in your first 2 paragraphs are totally fine and not the subject of this issue.

I'll try to state the problem differently. The current spec reads:

> When an operation requests to create a resource as a member of a container resource…

The point of this issue is that this specific "when" is expensive. A WAC system cannot tell for a given `PUT` or ` PATCH` request whether it is a creation by looking at the request itself. It has to go to the backend and check whether that resource exists.

Will add this to the issue text above.

**Comment by RubenVerborgh** [comment]
Clarifications after Solid Specification call on 2020-08-17:

I believe the ideal situation for scalability purposes would be as follows:
- WAC can be implemented as a layer that either:
  - returns `403` OR
  - sends the request through to the (LDP) backend, which then generates the status code (which is not `403`)
- This means that, to make a decision, WAC only needs as input:
  - the request
  - the ACL state
- Specifically, WAC does _not_ need any non-ACL Pod document state as input.

From this follows that (only) the following table states are problematic from a scalability perspective:
- rows where the status code in the case of "document exists" and "document does not exist" are different, AND one of them is `403`
  - because for those rows to be implemented, the decision would need to be made based on state held in different systems

**Comment by justinwb** [comment]
> I believe the ideal situation for scalability purposes would be as follows:
> 
> * WAC can be implemented as a layer that either:
>   
>   * returns `403` OR
>   * sends the request through to the (LDP) backend, which then generates the status code (which is not `403`)
> * This means that, to make a decision, WAC only needs as input:
>   
>   * the request
>   * the ACL state
> * Specifically, WAC does _not_ need any non-ACL Pod document state as input.

+1 - I believe that this decoupling is essential to scale and to good implementation design.

Also, would like to mention the value of having explicit modes for create / delete on a container in this context, as detailed in these authorization use cases (and associated requirements):

- https://solid.github.io/authorization-panel/authorization-ucr/#collection-readcreate
- https://solid.github.io/authorization-panel/authorization-ucr/#collection-readcreatedelete
- https://solid.github.io/authorization-panel/authorization-ucr/#collection-createonly

Having this added specificity in permissions would allow the authorization subsystem to make better and more accurate choices about what is does / doesn't pass through, while at the same time applying principle of least privilege to operations that could/would update the non-server-managed graph of the container.

**Comment by kjetilk** [comment]
I am very sympathetic to the goal of this issue. I will just record an opinion that I also voiced in the Solid Editor's meeting today:

I had a design for an auth proxy that could sit in front of the rest of the interaction, but in my design, I had an optimization that would record creates and deletions with the intention of keeping track of the existence of a resource (specifically into an in-memory trie). Originally, I didn't have this in mind, but simply to make it very cheap to reject a request with a `404` or `410`, if the resource did not exist, as I see this as a protection against a DoS attack that seeks to overwhelm the relatively expensive auth operations on non-existent resources. With this design, the above wouldn't be so much of a concern. Then, I feel that this is something that should be done anyway, and so, I don't see quite the importance of this issue.

Nevertheless, I acknowledge the concern, and that more access modes would be helpful, and that it should be something that we consider. Urgently, even.

**Comment by TallTed** [comment]
@RubenVerborgh --
> Clarifications after Solid Specification call on 2020-08-17:

I think that should probably say 2021, not 2020?

**Comment by csarven** [comment]
The tables in that comment are (obviously) populated from the point of the Protocol, not WAC's response codes.

PUT- and PATCH-to-create have the same requirements. I brought back the C/ column to the PATCH table and added the note: https://github.com/solid/specification/issues/14#issuecomment-683480525 . I've also added a row including `DELETE + INSERT` payload that we should review:

>|-|Read,Write|DELETE+INSERT|false|?|?|

"false" if no match for `DELETE`.

---

WAC ED says that server manages the association between resources and ACL resources so that applicable Authorization rules can be located (after effective-acl-resource) and evaluated. Let's pick on that: the backend, the WAC system, or something else needs to track that association. Conceptually similar to the "auth proxy" that Kjetil mentioned above but without tracking creates and deletes.

https://github.com/solid/web-access-control-tests/issues/41#issuecomment-898295543 :

>Typically there is no ACL resource for a non-existing resource, but the WAC spec doesn't prohibit that. (If there is an ACL resource for a non-existing resource, server would apply that ACL resource and checks for accessTo).

I'm curious to know how Solid servers + WAC implementations currently determine the effective-acl-resource.

---

>When an operation requests to create a resource as a member of a container resource

How can the server know an HTTP request strictly entails a Create operation? Here are some examples (there may be others):

|Request|Operation|
|-|-|
|`POST /C/`|Create|
|`PUT /C/R` + `If-None-Match`|Create, Read|
|`PATCH /C/R` + `If-None-Match` + `INSERT` in message body|Create, Read|

For `PUT` or `PATCH` (without `If-None-Match`), .. let's look at this right after the table.

Using CRUD terms in the table below for simplicity and the purpose of this discussion - do not bother to nitpick on this. Happy to switch to RFC terms or introduce 50 very specific terms. The WAC spec loosely uses the same terms for common operations. Yes, we can use "modify" or "replace" instead of "update", however that's not the point here.

* Important to not conflate resource-level e.g. create or delete resource, and content-level operations, e.g. add or remove statements.
* A request may entail one or more operations, including a processing operation depending on target resource that breaks down to specific operations on one or more resources.

Expect updates to this first pass at the table. Suggest changes/corrections. I'm not sure what to do with this table yet / where to put it but we can figure something out alongside https://github.com/solid/specification/issues/14#issuecomment-683480525 (even if it just serves to catch bugs or our assumptions)

|Request|Operation|Access Modes|Level|
|-|-|-|-|
|`GET /C/R`|Read|`acl:Read`|Resource|
|`HEAD /C/R`|Read|`acl:Read`|Resource|
|`OPTIONS /C/R`|Read|`acl:Read`|Resource|
|`POST /C/`|Create + Update|`acl:Append` [1]|Resource, Content|
|`PUT /C/R` + `If-None-Match`|Create, Read|`acl:Write`, `acl:Read` [2]|Resource|
|`PATCH /C/R` + `If-None-Match` + `INSERT`|Create, Read|`acl:Write`, `acl:Read`|Resource, Content|
|`POST /C/R`|Update|`acl:Append` (or `acl:Write`)|Content|
|`PUT /C/R`|Create?, Update|`acl:Write` [2][3]|Resource|
|`PATCH /C/R` + `INSERT`|Create?, Update|`acl:Write`, `acl:Read` [2][3]|Resource?, Content|
|`PATCH /C/R` + `DELETE + INSERT`|Create?, Read, Update|`acl:Write`, `acl:Read`|Resource?, Content [4]|
|`PATCH /C/R` + `DELETE`|Update, Read|`acl:Write`, `acl:Read`|Content|
|`DELETE /C/`|Delete|`acl:Write`, `acl:Read`|Resource|
|`DELETE /C/R`|Delete|`acl:Write` [5]|Resource|

[1] `POST C/` is to "perform resource-specific processing on the request payload" (RFC 7231). The semantics in Solid Protocol: server-assigned name for resource to be created as member of container, and to update containment statements. So, Create operation of `/C/R` and an Update operation on `/C/`. In WAC, Append is required on `C/`.

[2] Requires Append (or Write) on `/C/` and Write on `/C/R` to Create.

[3] Requires Write on `/C/R` to Update.

[4] `PATCH /C/R` + `DELETE + INSERT` could be interpreted as a specific content-level update (or modify) operation where DELETE s1p1o1 and INSERT s1p1o2 - I don't know what the practice/terminology is out there for this. We can come back to this later.

[5] Requires Write on `/C/` and `/C/R`.

---

>A WAC system cannot tell for a given PUT or PATCH request whether it is a creation by looking at the request itself. It has to go to the backend and check whether that resource exists.

WAC doesn't need to. Server does (as per tables above). WAC is asked about [2] and [3]. That can be one or more queries.

>WAC can be implemented as a layer that either:
>    returns 403 OR

>This means that, to make a decision, WAC only needs as input:
>    the request

Not strictly. For example - just to illustrate an idea for a possible implementation - the backend can send data about the collection hierarchy (or however collection hierarchy is determined) and the associated ACL resources. The WAC system somehow needs to know how a resource and its containers are connected, whether it is for example informed about slash semantics, checks for reverse relation from resource to its container, or something else. It also needs to be informed about (or keeps track of it) how resources and ACL resources are associated, which is required to determine the effective-acl-resource.

For `PUT /foo/bar/baz`, WAC in context of Solid will/can get to know the following paths and the associated ACL resources (if any):

```
/
/foo/
/foo/bar/
/foo/bar/baz

/.acl
/foo/.acl
/foo/bar/.acl
/foo/bar/baz.acl
```

WAC goes through the authorization-process, responds to the backend with allowed or denied:

For `PUT`, it checks for Authorization rule with Write access on `/foo/bar/baz` and Append (or Write) access on closest container (heading towards the root container) with an associated ACL resource that has an Authorization rule with said permission.

If allowed, backend determines if resource state is to be created or replaced, as well as any intermediate containers that needs to be created.

**Comment by RubenVerborgh** [comment]
@csarven Thanks for the detailed answer. However, I'm not sure if the core issue is addressed, so I will try to phrase it as a question:

**Do cases exist where WAC behavior depends on server (non-ACL) resource state?**

Because the "entails a Create operation" table above is (as you wrote) non-exhaustive: there are cases where a create operation will happen, but this cannot be seen in the request. Concretely, all cases in the updated https://github.com/solid/specification/issues/14#issuecomment-683480525 where exactly one column is a `403`, remain problematic.

**Comment by csarven** [comment]
Again, the final status codes are determined by the server. For GET /C/R, the backend must:

1: Check WAC for Read on /C/R.

1a: If there is a matching Authorization rule (=server can check storage):
1ai: If /C/R exists, return 200, otherwise return 404.
1aj: If /C/R does not exist, return 404.

1b: If there is no matching Authorization rule (=server does not need to check storage):
1bi: Return 403.

The backend may consider checking the following (in the same query with 1 above) as an alternative to 1b above:

2: Check WAC for Read on /C/:

2a: If there is a matching Authorization rule (=server does not need to check storage):
2ai: Return 404.

2a: If there is no matching Authorization rule (=server does not need to check storage):
2ai: Return 403.

Single query (snippet) below -- you'll have to excuse me as I'm probably getting the query wrong... it is Sunday night and I'm having sage tea, my own crop, fyi:

```
SELECT ?authorizationA ?authorizationB
WHERE {
  ?authorizationA
    acl:mode acl:Read ;
    acl:accessTo </C/R> ;
    acl:agent foo:#bar .

  OPTIONAL {
    ?authorizationB
      acl:mode acl:Read ;
      acl:accessTo </C/> ;
      acl:agent foo:#bar .
  }
}
```

Before I think through the other table rows, does the above bring us closer to a possible solution/perspective?

I said one or more queries because I think a single ASK or a SELECT query (with desired cols, rows) will do, but since it is an implementation detail, multiple ASKs or SELECTs can also work.

Marked with "Create?" because at some point the storage needs to be checked. It is not on WAC to know if resource exists. While I agree on the architecture to check access control before visiting the storage, I don't think an implementation choosing or perhaps even needing to visit the storage before and during with multiple trips makes it wrong - probably undesirable, inefficient, and things can potentially go wrong but it is still an implementation detail as far as the specs are concerned. Based on my understanding... but happy to be corrected on that or if there is an assumption somewhere that I'm not considering at the moment.

**Comment by RubenVerborgh** [comment]
> Again, the final status codes are determined by the server.

Full ack: I confirm I understand that the final status codes are determined by the server.

> Before I think through the other table rows, does the above bring us closer to a possible solution/perspective?

Not really; we seem to be working with different mental models of the systems.

My mental model is a pipeline in which a request goes through the *sequential* stages:
1. **WAC**: either returns 401/403 (skipping the next stage) or passes on to the next stage
2. LDP: serves documents and returns status codes such as 2xx, 400, 404

So once a request has left WAC stage, it cannot go back.

To check that we are on the same line here: do you see how the layered system I am sketching here (and is documented in more detail at https://github.com/solid/solid-architecture/blob/master/server/request-flow.md) has a separate place/stage where 403 is returned, and that this 403 decision is independent of whatever stage follows it?

--

You seem to be assuming a architecture in which some subsystem is asking questions to an ACL store and an LDP store.

So the system you call "WAC" seems to be an ACL store to which you can ask questions; the system I call "WAC" is a processing stage which makes a decision: 403 or PASS.

--

> it is still an implementation detail as far as the specs are concerned

It is not an implementation detail if the _only_ way to implement this is a slow and coupled way.

It also makes WAC not/less orthogonal of LDP. With a concrete example, take table row

```
GET C/R
HEAD C/R
OPTIONS C/R
```

|C/|C/R|C/R exists|C/R doesn't exist|
|-|-|-|-|
|Read|Write|403|404|

For the WAC subsystem to decide whether it should return `403`, it first needs to check whether the LDP system _would_ have given a `404`. We're instead interested to have a WAC subsystem whose decision for `403` is not influenced by LDP.

**Comment by csarven** [comment]
>My mental model is a pipeline in which a request goes through the sequential stages:
>So once a request has left WAC stage, it cannot go back.

I'm saying that's a fine implementation detail. I don't quite see how the spec prevents that possibility.

>You seem to be assuming an architecture in which some subsystem is asking questions to an ACL store and an LDP store.
>So the system you call "WAC" seems to be an ACL store to which you can ask questions; the system I call "WAC" is a processing stage which makes a decision: 401, 403, or PASS.

Can you break it down for me as to what's wrong with the pseudocode I wrote above? I thought it showed exactly what you wanted. Check if required Authorization rule exists, and check storage only when need to.

>For the WAC subsystem to decide whether it should return 403, it first needs to check whether the LDP system would have given a 404. We're instead interested to have a WAC subsystem whose decision for 403 is not influenced by LDP.

I think that's a misinterpretation of the pseudocode above.

**Comment by acoburn** [comment]
> My mental model is a pipeline in which a request goes through the sequential stages:
>   WAC: either returns 401/403 (skipping the next stage) or passes on to the next stage
>   LDP: serves documents and returns status codes such as 2xx, 400, 404

I approach this slightly differently.

The key point I would make is that the WAC layer would _not_ be responsible for producing a 403. (it may generate a 401)

Rather, the WAC layer would be responsible only for identifying what operations a particular agent _can_ perform on a particular resource. This stage might produce an array: `[read, write, append]`. But these do not need to correspond to the WAC modes. More likely they would be `[read, create, write, append, delete]` or `[read, create, append]` or a null set: `[]`, indicating that the agent has no access. Either way, this is all internal impl detail: only the Pod server needs to understand it. Importantly, this is where you need to define things like: can the agent create a new resource (e.g. `create`) or merely add triples to an existing resource (e.g. `append`). Can the agent delete the resource?

I would emphasize that the data at this stage is purely implementation detail.

Once that data arrives at the pod server / HTTP layer, you can inspect conditional headers (`If-Match`, etc) or whether the intended resource exists or not (is this a create or update operation?). That is when the 403 would be returned on error or 200/201/204/etc on success.

This structure will work with a completely externalized authorization server or with a single server that uses layers, as described above.

**Comment by kjetilk** [comment]
I think @RubenVerborgh and I share a mental model on this, since we worked this out in the early F2F while designing a NSS replacement. I think your design is very interesting, @acoburn , the reason why I/we (or whoever, don't remember) didn't further entertain that model was that it was perceived as a heavier than the approach taken by [acl-check.js](https://github.com/solid/acl-check/), which would return a boolean "authorized or not". 

Of course, there has always been some tension for our idea too, in particular around `PATCH` where the request body would have to be parsed. In your design, @acoburn , the authz evaluation could sit entirely within the authz component, and still let subsequent layers return the actual status codes.

In practical terms, how much of a burden do you think this would put on other implementers, @acoburn ?

**Comment by acoburn** [comment]
> In practical terms, how much of a burden do you think this would put on other implementers?

@kjetilk it is hard to say unequivocally whether a certain pattern puts too much burden on implementers. I _would_ say that the model described by you and @RubenVerborgh has the advantage of being very straight-forward (I've implemented variations on that model numerous times, and it is certainly the one that is more obvious).

One additional consideration, though, is the requirement for a `WAC-Allow` response header. For that, a server will need to know the modes for a given user (and the public user), which may require additional checks on the ACL layer (after a boolean access/no-access has been determined). By computing that access data once and then passing it down to the Pod layer, a server can avoid that additional check. Of course, if the impl translates ACL vocabulary terms into an internal data structure (e.g. `[read, create, append]`) that would need to be converted back to the standard WAC-Allow terms.

In summary, I would say that there are advantages and disadvantages to both approaches. In [my comment](https://github.com/solid/web-access-control-spec/issues/97#issuecomment-913227652), I was mostly trying to point out that it is _possible_ to construct a system where multiple round-trips are not required. Whether that model is desirable for a given implementation is very much dependent on context.

**Comment by RubenVerborgh** [comment]
Thanks @acoburn for shedding light on this!

In relation to `WAC-Allow`: in our design, the authorization layer adds that metadata.
The LDP layer just does not need to parse or interpret it, so it remains straightforward.

---

@csarven I will try to fully answer your concern:

> I don't quite see how the spec prevents that possibility.

In order to do so, I will provide a detailed breakdown of a possible strategy for request processing, followed by an explanation of where it breaks down. That why, I hope to demonstrate that the spec currently prevents this possibility.

### Reductive request processing for Solid

The idea of reductive request processing is that:
1. each incoming request goes through a **pipeline of stages**
2. each stage **reduces the complexity of the request** by taking care of one aspect of it
3. requests go through **each stage exactly once**
4. there is **no communication from one stage to previous stages**

Essentially, the HTTP request is seen as a set of problems that need to be solved, and every layer solves exactly one problem. That enables a high level of independence and a clean separation in codebases.

Here is a simplified version of the request processing pipeline.

0. **HTTP layer** receives an HTTP request.
    - This request contains all "problems" to be solved.
1. Next, the **authentication layer** extracts the agent's identify from the request.
    - From this point onward, the authentication parts of the request are _not_ accessed anymore. They are a solved problem.
2. Next, the **authorization layer** blocks the request if the agent does not have proper authorization.
    - It returns `401`/`403` if the user agent does not have permission to perform the requested operation.
    - From this point onward, the authorization parts of the request are _not_ accessed anymore.
    - In fact, _none_ of the layers below have to know _anything_ about authorization or authentication. That problem is considered solved at this stage.
3. Next, the **LDP layer** performs the requested operation.
    - It return `2xx`, `400`, `404` as needed by the operation.

### Why reductive request processing is important
Reductive request processing allows us to keep code focused on single concerns. The LDP layer does not need to know anything about authorization. The authentication layer does not need to know about LDP. This means that the LDP layer can be combined with any other authorization mechanism. It means that the WAC layer can be used with any other underlying API.

### Where it breaks
Reductive request processing works great for Solid. It has worked for WAC the way it was drafted in earlier versions. In fact, it works perfectly for 98% of possible cases of the current WAC ED… except for the 3-or-so cases in the tables where the status code depends on resource state. Those cases would require completely changing the architecture.

The reason it breaks is that those cases have a different status code depending on whether a request performs a creation. And that information cannot be conclusively deduced from the request itself, it requires accessing the resource state. Simply said: layer 2 breaks, because it cannot make a decision by itself. It has to contact other layers to make that decisions.

And it _could_. But below, I will argue why I think it _should not_.

By this point, I hope to have answered @csarven's question as to why the spec prevents this option. Please let me know if I have not made this point clear—and please read on to see why I think it is a problem.

### Why I think it should not break
We could make the argument that the above architectural model is fundamentally broken because it does not work with the WAC ED. And there are alternatives: a non-scalable monolithic approach (as in NSS) or an approach where the WAC layer instead passes information to an auth-aware LDP layer. While those are indeed _possible_, but there are reasons for not mandating those approaches.

The current ED forces one of two approaches:
1. either roundtrips from WAC to the LDP system are required
2. or the LDP system needs to have knowledge about authorization
The first approach leads to scalability issues, as discussed above. The second approach leads to a more complex LDP system.

Consequently, the ED does not allow for an independent WAC gatekeeping HTTP proxy; because that proxy would need a communication channel to the resource state.

It also removes the possibility to implement the Solid Protocol part of the specifications without being aware of access control, so limiting the orthogonality of specifications. (Unless the less scalable route of roundtrips is chosen, but then you have that problem.)

### Disclaimers

> I'm saying that's a fine implementation detail.

I hope that I have made the case that the current ED rules out _all_ such simple implementations, given that it only allows two approaches (roundtrips or auth-aware LDP).

### What I am asking for
Let's avoid cases where resource state is required to determine whether the status code should be `403`.
Consequently, the request itself (and the applicable ACL document) should contain all the data to make that decision.

We can achieve that by either:
1) ensure that the request itself unambiguously states whether it is a creation request (that would be a change to the Solid protocol)
2) make the status codes for "Create?" cases the same, independently of whether the answer turns out to be "yes" or "no"

I do not think that 1 is desirable, so that seems to leave us with 2.
Note that an acceptable solution for 2 could be that either a) `Append` on the container is required for both creation and update or b) `Append` on the container is required neither for creation nor for update. In other words, I don't care what the exact behavior is, as long as the behavior is independent of the answer to the "Create?" question.

### Summary
Only 3 specific cases out of dozens prevent implementers from choosing a straightforward and elegant layer-based model. I think the more logical fix is to adjust the cases, rather than mandating that every single implementation follows a more complex model.

**Comment by csarven** [comment]
One or both of us is not understanding the other. Let me re-quote from earlier because I think it is still a misunderstanding and I want us to be on the same page on this:

>For the WAC subsystem to decide whether it should return 403, it first needs to check whether the LDP system would have given a 404.

No. WAC strictly decides what's allowed and/or denied based on what's being queried. When the system that gets an answer from WAC, it proceeds by deciding whether the storage needs to be checked or not, and then decides on the status code.

>1: Check WAC for Read on /C/R.

Done with the Authorization layer.

>1a: If there is a matching Authorization rule (=server can check storage):

Access is allowed so use the storage.

>1ai: If /C/R exists, return 200, otherwise return 404.

Done.

>1b: If there is no matching Authorization rule (=server does not need to check storage):

Access is denied. There is no need to use the storage.

>1bi: Return 403.

Done.

---

Now, let me respond to this because alternatively a query to WAC can check for permissions on /C/R and /C/ (that was the rough SELECT example above). This is entirely optional. It is just a way to be smart/realistic about the response code.

>2: Check WAC for Read on /C/:

Done with the Authorization layer.

>2a: If there is a matching Authorization rule (=server does not need to check storage):

Access is allowed i.e., /C/R can be seen in /C/. There is no need to use the storage.

>2ai: Return 404.

Done.

>2b: If there is no matching Authorization rule (=server does not need to check storage):

Access is denied i.e., /C/ can not be read. There is no need to use the storage.

>2bi: Return 403.

Done.

---

Is there anything you disagree with the above or that's not clear?

**Comment by RubenVerborgh** [comment]
> One or both of us is not understanding the other.

I think that's because you're replying to an old comment on this thread, which I have since clarified.
Unless I am mistaken, I have not seen an acknowledgement or follow-up on the clarification.

> > For the WAC subsystem to decide whether it should return 403, it first needs to check whether the LDP system would have given a 404.
> 
> No. WAC strictly decides what's allowed and/or denied based on what's being queried.

When I write "the WAC subsystem", I am not talking about what you call "WAC", but rather about the authorization layer.
To clarify this, I had written [this comment](https://github.com/solid/web-access-control-spec/issues/97#issuecomment-913851982) where I use separate terms for those things.

Could you please start from that comment to have this discussion?
Otherwise, we will likely keep on being confused.

**Comment by kjetilk** [comment]
You know, I think this is whiteboard matter stuff... We aren't communicating well enough because our medium doesn't have the emotional and conversational bandwidth we need. Everyone should keep that in mind, and so, we try our best.

So, that being said, I just wonder if the problem arises simply by that there is a Create operation? I prefer to think in terms of operations, and then there are access modes that allow certain operations to be performed.

Take the case of a `PUT`, which may be either a Create or a Modify operation. You can ensure that it is a Create operation by using the `If-None-Match` header, but you don't have to, so in principle, when the server sees a `PUT` it can be either, you can't tell for sure unless you know the state of the target resource.  

Now, say that we have a create access mode, and the client sends a `PUT` request and is authorized with that create access mode, but not modify. Then, the authz layer would have to know the state of the resource to determine the result. 

I very much like a layered architecture, but I thought about it in slightly different terms. To me, it was OK to have the layers have bi-directional communication if the communication happened in the normal request-response cycle (thus calling it a proxy). I had another restriction though, and that was that the request body should only be parsed once (so the `PATCH` "handler" would have to deal with authz too).

If you allow for the response to be seen by the top layers, then you can record the state of the resource. Not only whether it exists, but you can also have mtime and etags recorded in the top layer, lock a resource representation to deal with concurrent requests and so on.

So, that seems to me as an alternative architecture. 

If I was writing a server now, I would probably adopt @acoburn 's idea of having the second layer determine the allowed access modes, and let that communicate them down. At the bottom, I would then code based on operations and have methods that checked the access modes for each operation. Then, the lowermost layer would not actually have to evaluate the authorization that was done in the second layer, but it would need to know them, but that seems like a small price to pay.

**Comment by csarven** [comment]
Ruben, is the architecture you're referring to reserves 403 to the authorization layer?

What's (currently) included in that processing stage besides an access control system?

Would any feature that could potentially conclude 403 is to be included in the authorization layer?

---

"403,404" (or even "200,201") is discussed from the point of the Protocol: 403 is a ~MUST whereas 404 would be a ~MAY.

403 could be decided based on the evaluation of an access control system but it is not necessarily confined to it (or a particular architecture's authorization layer). For example, the protocol or even a particular server may disallow allocating certain URIs or pattern without needing to check actor's permissions.

I think this concern is not specific to WAC, or any other access control or capability-based security -- unless they wilfully violate or intended to override protocol's response codes.

I propose that we consider discussing this issue from the point of the solid/specification. Related: https://github.com/solid/specification/issues/288 , https://github.com/solid/specification/issues/146 , https://github.com/solid/specification/issues/219#issuecomment-876405173 .

---

I'm sensing that you want the server to be able to return a 404 but that would mean that the 403 possibility is still there in case when agent doesn't have Read on container. But the architecture seemed to reserve 403 to the authorization layer - after all, if it PASSed the authorization layer, 403 should not arise from the following stages. If 403 is still possible in stages after (or even before) the authorization layer, then it can be used failing the check for 404 - this is what I've been trying to communicate in that, if you want the 404 possibility, you need to check storage [1].

Is this an accurate summary?

One way out of this is to not bother with the 404 possibility obviously, but let's continue..

---

>ensure that the request itself unambiguously states whether it is a creation request (that would be a change to the Solid protocol)

Right. We know for some cases but not all - the table in https://github.com/solid/web-access-control-spec/issues/85#issuecomment-913456115 distinguishes some.

>I do not think that 1 is desirable, so that seems to leave us with 2.

Right, we are not going in that direction at this point in time. FWIW, alternative approaches, e.g., POST with Activity Streams 2.0 activities such as Create, Delete, Add, Remove.

>make the status codes for "Create?" cases the same, independently of whether the answer turns out to be "yes" or "no"

That would prohibit implementations capable of enhancing their product by providing specific/useful codes.

---

I find it strange that the differences in status codes are potentially an issue for some architectures - as you seem to be suggesting - but what's in the table is not above and beyond vanilla RFC 7231.

---

>an acceptable solution for 2 could be that either a) Append on the container is required for both creation and update

For creation, that's possible (having Append (or Write) on /C/ is required in order to create /C/R, so that's possible.) However, for updating, I don't think Append on /C/ should be required if only Append is desired on /C/R.

>or b) Append on the container is required neither for creation nor for update.

Append on /C/ is useful to allow processing operations entailing POST (as per the resource semantics of C/) and to have the server to allocate the URI of a resource (creating) that's to be a member of /C/. As for updating /C/R, possible to just Append on it.

---

Re Kjetil's:

>Now, say that we have a create access mode, and the client sends a PUT request and is authorized with that create access mode, but not modify. Then, the authz layer would have to know the state of the resource to determine the result.

Exactly. This also shows why throwing in new access modes doesn't address the concern at hand.

**Comment by RubenVerborgh** [comment]
Thanks, Sarven.

> Ruben, is the architecture you're referring to reserves 403 to the authorization layer?

Yes; slight clarification: in the proposed architecture, all `403`s thrown _for WAC reasons_ are thrown b the authorization layer.

> What's (currently) included in that processing stage besides an access control system?

Nothing. (Authentication happens in the layer above it.)

> Would any feature that could potentially conclude 403 is to be included in the authorization layer?

- If the reason is WAC, always yes.
- If not WAC, likely not (but possibly).

----

> 403 could be decided based on the evaluation of an access control system but it is not necessarily confined to it (or a particular architecture's authorization layer). For example, the protocol or even a particular server may disallow allocating certain URIs or pattern without needing to check actor's permissions.

That's all fine; other reasons can still throw `403`, if not for WAC reasons.

For example, in CSS, it is _another_ layer that throws the `403` for a particular clause of the Solid protocol https://solid.github.io/specification/protocol#writing-resources (update auxiliary resource with `Slug`).

>  But the architecture seemed to reserve 403 to the authorization layer - after all, if it PASSed the authorization layer, 403 should not arise from the following stages.
> […]
> Is this an accurate summary?

----

Not fully. The accurate summary is: all `403` responses _mandated by WAC_ should be able to be generated by a process that only has the HTTP request as input, without further communication with other parts of the system (storage in particular).

> > make the status codes for "Create?" cases the same, independently of whether the answer turns out to be "yes" or "no"
> 
> That would prohibit implementations capable of enhancing their product by providing specific/useful codes.

No; those implementations can choose to follow another architecture.

----

> I find it strange that the differences in status codes are potentially an issue for some architectures - as you seem to be suggesting - but what's in the table is not above and beyond vanilla RFC 7231.

That's not strange at all: RFC 7231 is very wide spec. SOAP+WSDL also does not go beyond RFC 7231, but it is extremely hard to scale because of the way it uses that spec. A large subset of RFC 7231 does not scale at all; some of its subsets have very beneficial properties. With the Solid protocol being a subset of RFC 7231, the question is what properties we want to pick.

Another example to open our minds:
- The regular expression syntax is quite powerful.
- However, if we stick to a specific subset of regex, we can implement processors that have linear time complexity.
- If we stick to a subset where 95% of expressions are simple, but 5% are not, we still need a more complex processor to handle the 100%.

So what I'm saying is: 95% of WAC currently sticks to a subset of RFC 7231 that has beneficial properties from a scaling/complexity perspective. The question is thus: do we want, for those 5% of cases, to make the entire system more complex, or do we want to adjust those cases such that 100% of WAC stays within the RFC 7231 subset with beneficial properties?

----

> > an acceptable solution for 2 could be that either a) Append on the container is required for both creation and update
> 
> For creation, that's possible (having Append (or Write) on /C/ is required in order to create /C/R, so that's possible.)

That would be my preference.

**Comment by RubenVerborgh** [comment]
Another line of argument: if you currently want to implement new storage for the Community Solid Server, you don't need to know about WAC.

We have projects that add support for IPFS and Phillips Hue lights. They work out of the box, no access control knowledge needed.

----

If the status code rules as written in that comment are to be followed, then people implementing storage for CSS also need to know about access control, making it harder to write such modules.

And you could say: well then CSS has to change the way it works. But the argument above shows that we cannot. Because of the current status code choice, _all_ implementations either have scalability challenges (NSS) or force storage developers to know about access control (ESS). We've proven above that there is no other way.

So those choice make implementations harder; still feasible, but we'd like to avoid it if we can.

**Comment by kjetilk** [comment]
@RubenVerborgh , I agree with you in principle but not in detail.

I agree that it is important that the spec makes possible architectures with good quality attributes and encourages implementations of such. 

However, when you say that:
>  But the argument above shows that we cannot. Because of the currently listed status code choices, all current and future implementations will either have scalability challenges (NSS) or force storage developers to know about access control (ESS). We've proven above that there is no other way.

You have proven no such thing. You haven't responded to my proposal of a reverse proxy, which would resolve the issue, and would be minor modification both to the architecture and implementation. 

I might be old, but I would seriously never consider going into production without a reverse proxy, even on my personal box. It appears that is also best practice for [nodejs](https://www.nginx.com/blog/5-performance-tips-for-node-js-applications/) and [express](https://expressjs.com/en/advanced/best-practice-performance.html#use-a-reverse-proxy). It also looks simple [to get started](https://www.twilio.com/blog/node-js-proxy-server). I just don't see why you would resist this idea.

So, you can keep the stacked architecture and the reductive processing, you just need to have a reverse proxy that tracks the existence of a resource, and let that proxy do the first reductive step (i.e. return a `404` (should also do `304` etc)). It will also communicate whether the resource existed for requests like `PUT`. The rest remains unchanged. With this, you don't need to roundtrip to the backend.

The key problem with your current architecture is that it cannot support create operations with the MUSTs of RFC7231, and it really has to.

I suggest that you make a simple node-based reverse proxy. We should also implement a library in C that can be integrated in stuff like nginx, HAproxy or Varnish at some point. The reverse proxy could also be offloaded to a CDN. I think the benefits of this architectural change is great, much greater than the downsides.

BTW, I just found me architecture brain dump from waaaaay, back.
![next-gen-solid-server](https://user-images.githubusercontent.com/154068/134251709-230a80b6-3350-4a0a-95c5-76e630165e1b.png)
 
The "Resource list cache" is that architectural component. I still think it is required.

**Comment by RubenVerborgh** [comment]
> you just need to have a reverse proxy that tracks the existence of a resource

This assumes that there is only one interface to the data, quod non.

Examples where this breaks:

- file-based backend, files are added directly to the filesystem
- sensor-based backend, new observations are created by the sensor
- the server has LDP, SPARQL and GraphQL interfaces
- the server exposes the same data through multiple virtual folders via LDP

So I rest my case that we have to choose between scalability challenges or a ACL-aware storage.

**Comment by kjetilk** [comment]
> > I would seriously never consider going into production without a reverse proxy
> 
> Me neither; CSS has explicit support for this.
> 
> > you just need to have a reverse proxy that tracks the existence of a resource
> 
> This assumes that there would be only one interface to the data, quod non.

Well, obviously, if you add interfaces, you should be prepared the tackle the consequences, including authentication, authorization, 
concurrency control, etc. I don't see it as appropriate to tackle the requirements of all possible different interfaces, even though I certainly see them coming in the IoT space, and where many of them aren't even neatly layered.
 
> Examples where this breaks:

I would not say that, I would say "Examples where I may choose to break this". Now, it can be done, but what if we add end-to-end encryption to the protocol, which is certainly an important feature?  

That said, it isn't usually hard:

>     * file-based backend, files are added directly to the filesystem
>     * sensor-based backend, new observations are created by the sensor

Then, the implementation will have to notify the front proxy.


>     * the server has LDP, SPARQL and GraphQL interfaces
>     * the server exposes the same data through multiple virtual folders via LDP

They should be going through the same proxy on the way out.



> 
> So I rest my case that we have to choose between scalability challenges or a ACL-aware storage, which is a real problem.
> Happy to put reverse proxies on top of pod servers, but that is an orthogonal issue to me.

I'm sorry that I wasn't present when you started writing CSS, because I could have commented at an early stage, I understand that you are pretty deeply tied to the architecture now. However, I do not think you have made a good case here, it is very clear that this is not a matter of what is possible or not, when it comes to this, it is a matter of trade-offs, and I think you have made a wrong trade-off, as one of the key reasons for using a reductive model is the ensure that you respond in the the cheapest way first. You do the opposite, you first do the auth, and so make the server susceptible to a trivial DoS attack. An attacker can simply flood the server with requests with random target URLs, and it will end up doing authz for all of them. That has to be prevented, trust me, I've been there, done that, and not been rewarded a t-shirt for it (well, it wasn't me, but it was my team). Returning a `404` cheaply is really your first line of defense. Having to go to storage for that will have very bad consequences. You'd have to mitigate with `fail2ban` or something like that on the front-end from very early on.

What is not a trade-off however, is that you can't support `PUT` properly.

**Comment by csarven** [comment]
>WAC ED says that server manages the association between resources and ACL resources so that applicable Authorization rules can be located (after effective-acl-resource) and evaluated.

Isn't the existence of a resource known?

**Comment by RubenVerborgh** [comment]
People, this issue is getting so much more complex and involved than it needs to be.

My original point is really simple: 3 rows in some table necessitate a more complex architecture. Do we want that, yes or no?

----

> I understand that you are pretty deeply tied to the architecture now

>  I think you have made a wrong trade-off,

That is __not__ the point at all. We can change CSS in a heartbeat.
Takes 2 hours of work after https://github.com/solid/community-server/pull/962.

The point is: I think it's a bad idea to do so.

> one of the key reasons for using a reductive model is the ensure that you respond in the the cheapest way first

Not our reason. Performance and DoS resilience are not CSS design goals.

Our design goal is a simple and modular implementation of the Solid specs, where implementers need the least amount of knowledge possible to run, configure, and build things.

----

Whether or not we use a proxy is 100% orthogonal to the question of whether we can make ACL decisions without needing resource state. The proposed resource list cache is (indirect) access to storage, so whatever algorithm is running in the cache/proxy needs to be storage-aware, just adding further evidence to my point above.

The proxy can be put in front of ESS, CSS, NSS, and it will make cheap decisions first. And that's great, let's build it!
I just want all 3 servers to also to be able to function without it. Requiring a proxy only reinforces the point that the current spec is complex to implement. I cannot ship a CSS that only works with a complex proxy+caching solution in front of it.

----

As the opener of this issue, the resolution I am asking for is either:
- **YES, it is a deliberate choice to make ACL`403` decisions require more than just the HTTP request, because of 3 rows in status code table.** We understand that this makes implementations/deployments more complex, just to support these 3 cases, and we are fine with that.
- **NO, let's not make ACL more complex just for 3 cases.** Let's adjust those cases such that  `403` decision for ACL reasons can be decided purely using the HTTP request.

In both cases, proxies can optionally be put in front of the server and make things faster.
In the second case, we can also keep the ACL implementation simple.

**Comment by csarven** [comment]
Ruben, we are trying to get answers just as you're! When I ask a simple yes or no question, I'd like a simple response as well. It is not always obvious what the answer is by working through a long response.

In any case, I've been revisiting some of the fundamentals. I'll have to get back to you on a longer response, but I believe the row for GET can be 403 only. (Even though I mentioned several times that the 404 path is completely optional.) The reasoning is that only the request target is relevant since the method is safe and idempotent.

**Comment by RubenVerborgh** [comment]
> Ruben, we are trying to get answers just as you're! When I ask a simple yes or no question, I'd like a simple response as well.

Thanks @csarven; sorry if I appeared impatient.

If the question you want an answer to is this one:

> > WAC ED says that server manages the association between resources and ACL resources so that applicable Authorization rules can be located (after effective-acl-resource) and evaluated.
> 
> Isn't the existence of a resource known?

The information can be obtained from the resource state storage indeed.

So is it "known": you'd have to access that system to make it known. That access might be cheap or expensive. But you need to do it.

**Comment by csarven** [comment]
>So is it "known": you'd have to access that system to make it known. That access might be cheap or expensive. But you need to do it.

Right, wherever that information is. My main point was that the knowledge of the association is necessary to apply the Authorization rules. And knowing that association also reveals the existence of a resource.

**Comment by RubenVerborgh** [comment]
> And knowing that association also reveals the existence of a resource.

Not necessarily. ACL documents and their corresponding resources might (and routinely do) live in different systems.

So if I ask the ACL document storage for `/a/b/c.acl` and I ultimately get `/a/.acl`, I don't know whether `/a/b/c` exists.

**Comment by csarven** [comment]
What's determining the effective-acl-resource?

**Comment by RubenVerborgh** [comment]
Different options; in the most general sense, there is an ACL Document Store that you can ask "give me the ACL triples that would apply to a resource named xyz".

**Comment by csarven** [comment]
We need to move this issue to solid/specifications. I'm okay to keep it here as we reach consensus but with the understanding that it is not a WAC bug. Any objections or preferences?

**Comment by RubenVerborgh** [comment]
Fine with me, thanks.

**Comment by dmitrizagidulin** [comment]
@RubenVerborgh - thanks for detailed writeup of the problem. And, I agree with you, getting the WAC spec to a point where it can be handled via the reductive architecture, would be very advantageous.

I'd love to hear from @kjetilk and @csarven with regards to - so what's _downside_ of refactoring WAC slightly (to remove those 3 exceptions)? Is it just the general friction of making a substantial change to the protocol? 

If I'm understanding the 3 cases correctly, we'd get a potentially serious boost in implementation simplicity and performance, in exchange for only minor protocol-change-pains.

**Comment by kjetilk** [comment]
@dmitrizagidulin I have decided to not comment any further on this until I have had a more high bandwidth conversation with @RubenVerborgh . The short story is that I do not think this is a WAC issue at all, the impact on future Solid is great, and that we need to take into account what kind of environment we are deploying into.

**Comment by elf-pavlik** [comment]
> A WAC system cannot tell for a given PUT or PATCH request whether it is a creation by looking at the request itself.

If we require that create always uses `If-None-Match: *` it would be possible, just based on that header, to tell if it is creattion.

**Comment by kjetilk** [comment]
> > A WAC system cannot tell for a given PUT or PATCH request whether it is a creation by looking at the request itself.
> 
> If we require that create always uses `If-None-Match: *` it would be possible, just based on that header, to tell if it is creattion.

Indeed. Do we want to do that? To me, it seems like it puts a big expectation on a client, but then again, we currently have just a write and append, so if we do that, it could be very helpful.

**Comment by csarven** [comment]
Revisiting the assumptions. Fun times.

Wrote this earlier ( https://github.com/solid/web-access-control-spec/issues/97#issuecomment-913190774 ) which I think clarifies the situation we are in:

> https://github.com/solid/web-access-control-tests/issues/41#issuecomment-898295543

>>    Typically there is no ACL resource for a non-existing resource, but the WAC spec doesn't prohibit that. (If there is an ACL resource for a non-existing resource, server would apply that ACL resource and checks for accessTo).

That clarifies everything right? Okay, one more moment..

Taking the following row as an example:

```
PUT C/R
```

|C/|C/R|C/R exists|C/R doesn't exist|
|-|-|-|-|
|-|Write|200|403|

I'll expand:

1. There is an Authorization rule for C/R and C/R exists.
2. There is an Authorization rule for C/R and C/R does not exist.

So, this depends on how a system is designed. It is not even a Protocol issue because we haven't imposed anyting beyond what RFC 7231 says! I've entered the codes as accurately as I can, and I trust that you've verified their accuracy - (I don't mean to imply that there are no other errors.)

Scenario 1 is a non-issue. Authorization is approved. Storage is checked, C/R exists, happy response.

Scenario 2 is an edge-case weird possibility, but still correct as per RFC. It shows that the system has (allowed!) an Authorization rule for C/R where C/R doesn't exist. The system got approval to write C/R. Storage is checked, C/R does not exist, sad response.

The system should either a) not have the request proceed with full range of allows or disallows for resources that do not exist, or b) not have Authorization rules for resources that do not exist.

The table row wasn't accidental. I wrote it fully knowing that possibility, where there is an authorization rule for something that does not exist. It was intended to demonstrate the status codes. Systems don't have to be designed to cover that case. They can work around it. But if they want to allow scenario 2 and conform to RFC, that's how it is. I think.

That's the best explanation I can give right now.

Computer, exit.
Computer, arch.

**Comment by RubenVerborgh** [comment]
> Wrote this earlier ( [#97 (comment)](https://github.com/solid/web-access-control-spec/issues/97#issuecomment-913190774) ) which I think clarifies the situation we are in:
> 
> > [solid/web-access-control-tests#41 (comment)](https://github.com/solid/web-access-control-tests/issues/41#issuecomment-898295543)
> 
> > > Typically there is no ACL resource for a non-existing resource

I don't think that's necessarily the situation we're in.
The ACL rule that applies to the resource can also be at the level of the folder.

So to me, the existence of a resource-specific ACL document is orthogonal to this issue.

---

> But if they want to allow scenario 2 and conform to RFC, that's how it is.

Understood. However, we are at liberty to change "how it is", which is what I'm suggesting.

Concretely: if `Write` permissions exist to a non-existing resource `C/R`, please let me also `PUT` or `PATCH` it into existence. There is no contradiction with RFC 7231 if we allow that; it's a choice on the level of WAC, as far as I understand.

**Comment by michielbdejong** [comment]
> Given that solid/specification#197 has been decided, I think it is a difficult thing to change though.

Thanks for clarifying @kjetilk! So I think this issue can be closed, then.

Let's conclude that we thank Ruben for submitting this proposal, but kindly ask the CSS team to align their implementation of WAC with the other existing implementations.

**Comment by RubenVerborgh** [comment]
@michielbdejong It is weird to read statements like the above, where you seem to assume some authoritative position over the community. Please don't do this; you have an equal say in matters, like everyone else. And if you want to help make decisions about the specification, I suggest you join the editorial team and deliver the due diligence that comes with it.

I don't need to be thanked or suggested to do anything; what I need is a technical solution to the coupling of authentication and storage backend. Technical input is welcome.

**Comment by michielbdejong** [comment]
No authoritative position, just stating that I think we (the community) should politely ask you (a valued member of our community) to (finally) align with the decision that we took as a group.

You opened https://github.com/solid/specification/issues/197 in June 2019, and last July it was closed after two years of debate. It has been decided, and all servers except CSS have aligned. Why is that?

As I commented on https://github.com/solid/community-server/issues/618 :

> permission handling does not take into account if a resource already exists or not.

That is still a bug, even if you delete my comments there.

**Comment by RubenVerborgh** [comment]
> Why is that?

Because contrary to your [claim above](https://github.com/solid/web-access-control-spec/issues/97#issuecomment-1000723479), this issue is a **bug report**, not a change proposal.
When this bug in the spec is solved—in whatever way—then it will be implemented.

Also, CSS implements v0.9 of the spec text regarding this matter.
I'd suggest the same for other servers and test suites, that would be awesome.

**Comment by csarven** [comment]
@RubenVerborgh , it was explained that this is not a WAC bug and the label was removed - something you've agreed with: https://github.com/solid/web-access-control-spec/issues/97#issuecomment-926524508 -  but I'm now noting that you've introduced the label without discussion.

**Comment by RubenVerborgh** [comment]
@csarven My apologies. Given that this coincided with a title change, I probably noticed no labels were assigned and added one without thinking.

That said, I wanted to get in touch with you about this issue. I think the long discussion has clouded the main point, but at the same time it has provided a lot of input that helps me clarify the issue.

Therefore, I would suggest that I make a new, more clearly phrased issue, incorporating the input above, such that it is easier to make progress. If you agree, what would be the best place to open it?

**Comment by timbl** [comment]
Happy for this discussion to be moved to another issue.   But while I am here I want to leave a should rant against false economies.  Solid works well when the protocol allows powerful operations.  If you shirk doing something like iterating over a path `/a/b/c/d/e/f` in the server, Remember how it is orders of magnitude slower to force the client to iterate over them as each iteration is across the global internet and is limited at the end of the day by the speed of light.   That that the server creates intermediate directories automatically with a PATCH INSERT to a non-existing file is great speed-up, for which no amount of internal server optimization can compete.   
If by giving insufficient data about container contents to the client forcing it to iterate over the folder contents doing a HEAD, that's massively slow.  If you you don't tell a client whether a potentially forbidden  exists forcing the client to get the user to log in and retry with credentials to find actually it does not exist, that's not an optimization. (Yes it might be more secure but it is dysfunctional).   So if giving better service to the client and the user involves running the Authorization function in parallel with the Storage function every time (for GET/HEAD) rather than starting the storage only of the auth gives an ok, maybe that makes solid a more useful platform.   To make the server faster but less powerful would maybe be a false economy.   For PATCH, clearly you can't do them in parallel, and maybe the corner cases are different, but still beware of false economies.

**Comment by timbl** [comment]
For comparison here is Google docs shows you a file which does not exist
<img width="752" alt="Screen Shot 2022-02-09 at 12 55 24" src="https://user-images.githubusercontent.com/1254848/153205582-068668be-773f-4daf-a91e-2389c81ae1b6.png">
:

compared to one which does exist but you don't have access to:
<img width="776" alt="Screen Shot 2022-02-09 at 12 55 47" src="https://user-images.githubusercontent.com/1254848/153205744-bd187298-1f52-41d6-bea4-03ba36515379.png">

The distinction is important.  People expect it I think. It does provide an onboarding path with the [protected document](https://docs.google.com/document/d/1C3rjT-aUi8aORiwi7tjIEzVHVjcDFCCJgYE1tKN8zjI/edit) , but it does not  waste user's time with the [one which does not exist](https://docs.google.com/document/d/blah1tKN8zjI0000000000000000000000000000000/edit).

**Comment by csarven** [comment]
@RubenVerborgh solid/specification
